### ISSUE TITLE
Enable Reweighing to handle categorical output instead of only 0/1 

### ISSUE THREAD

**Issue by adrinjalali** [fairlearn/fairlearn#911]
From: https://github.com/fairlearn/fairlearn/pull/906#discussion_r673118210

Some of our algorithms require the output to be 0/1. However, they only need the output to be categorical, and in case of binary classification they need to know which one is the positive label, i.e. `pos_lable`. This issue is to fix those instances.

**Comment by romanlutz** [comment]
We may want to separate this into several issues (one per algorithm) if it turns out to be a lot of work. I do expect there to be a lot of common foundational work that can be shared by all of them since they already share the input validation code.

**Comment by adrinjalali** [comment]
This issue certainly doesn't have to be fixed in a single PR, we could update the description and list the algorithms which are relevant and go through them one by one, or we could have a separate issue for each, I'm easy either way.

**Comment by SeanMcCarren** [comment]
Hi @adrinjalali @romanlutz, I think scikit-learn's `LabelEncoder` can be used? I could give it a try if this issue is still relevant :)

**Comment by romanlutz** [comment]
@SeanMcCarren it's definitely still relevant! I would suggest starting with one algorithm only and trying it out with LabelEncoder. Thanks for volunteering!

**Comment by SeanMcCarren** [comment]
I've delved a bit deeper into the issue now. Starting with a list of algorithms to update (as @adrinjalali mentioned):
- [ ] threshold optimizer
- [ ] exponentiated gradient
- [ ] grid search

Then, my thoughts so far... My initial thought was that we could just translate strings to numbers in the shared input validation and be done with it. But I now think that this does not work, because we need to persist this encoding throughout calls to fit & predict. 
I assume we should follow [this sklearn guideline](https://scikit-learn.org/stable/developers/develop.html#specific-models) which tells us that we should store a list of classes in `self.classes_` on every estimator. Seems clear!

Side notes: 
1) preferrably, I'd avoid writing the same piece of code to do this in every algorithm, but then I see that sklearn actually does repeat the initialization of `self.classes_` at times, instead of capturing this in some mixin or base class. I'd rather follow sklearn here then, but I wonder why sklearn made this decision (@adrinjalali maybe you know?)
2)  In `interpolated_thresholder.py` line 141 there is `enforce_binary_labels = False` while in `threshold_optimizer.py` line 288 there is `enforce_binary_labels = True`. I can not infer whether one of the two is an error or whether I can assume that threshold optimizer only works on binary data.

**Comment by adrinjalali** [comment]
@SeanMcCarren overusing Mixins makes the code less readable, and moves the codebase more towards a large framework kinda codebase, which is hard to maintain and read and change especially for new contributors. If the repeated code is just a single line of code, we don't try to abstract it away.

**Comment by taharallouche** [comment]
I'll try to work on this, starting with the `ThresholdOptimizer`.
